drivers/interrupt_controller/intc_gicv3.c

```diff
@@ -0,0 +1,319 @@
+#define IGROUPMODR(base, n)	(base + GIC_IGROUPMODR + (n) * 4)
+
+#define NUM_INTR_PER_REG	0x1F
+/* system register interface to GICv3 */
+#define ICC_IGRPEN1_EL1         S3_0_C12_C12_7
+#define ICC_SGI1R               S3_0_C12_C11_5
+#define ICC_SRE_EL1             S3_0_C12_C12_5
+#define ICC_SRE_EL2             S3_4_C12_C9_5
+#define ICC_SRE_EL3             S3_6_C12_C12_5
+#define ICC_CTLR_EL1            S3_0_C12_C12_4
+#define ICC_CTLR_EL3            S3_6_C12_C12_4
+#define ICC_PMR_EL1             S3_0_C4_C6_0
+#define ICC_RPR_EL1             S3_0_C12_C11_3
+#define ICC_IGRPEN1_EL3         S3_6_c12_c12_7
+#define ICC_IGRPEN0_EL1         S3_0_c12_c12_6
+#define ICC_HPPIR0_EL1          S3_0_c12_c8_2
+#define ICC_HPPIR1_EL1          S3_0_c12_c12_2
+#define ICC_IAR0_EL1            S3_0_c12_c8_0
+#define ICC_IAR1_EL1            S3_0_c12_c12_0
+#define ICC_EOIR0_EL1           S3_0_c12_c8_1
+#define ICC_EOIR1_EL1           S3_0_c12_c12_1
+#define ICC_SGI0R_EL1           S3_0_c12_c11_7
+
+#define	GIC_SPI_BASE		32
+
+#define GIC_INTID_PENDING_NONE	1023
+/* GIC idle priority */
+#define GIC_DEFAULT_PRIO	0xff
+
+/* Interrupt group definitions */
+#define ENABLE_G0		0
+#define ENABLE_G1NS		1
+#define ENABLE_G1S		2
+#define RWP			31
+/* ICC_SRE_ELx SRE bit */
+#define SRE			0
+#define DIF			1
+#define DIB			2
+
+/*
+ * wait for read write progress
+ * TODO: add timed wait
+ */
+static int gic_wait_rwp(u32_t intid)
+{
+	mem_addr_t base;
+
+	base = (intid < GIC_SPI_BASE) ? GICR_CTRL : GICD_CTRL;
+
+	while (sys_read32(base) & BIT(RWP))
+		;
+
+	return 0;
+}
+
+void arm_gic_irq_set_priority(unsigned int intid,
+			      unsigned int prio, u32_t flags)
+{
+	u32_t bit = intid & NUM_INTR_PER_REG;
+	u32_t mask = BIT(bit);
+	u32_t shift;
+	u32_t idx = intid / 32;
+	u32_t val;
+	mem_addr_t base;
+
+	base = (intid < GIC_SPI_BASE) ?
+	       (GIC_RDIST_BASE + GICR_SGIBASE_OFFSET)
+	       : GIC_DIST_BASE;
+
+	/* Disable the interrupt */
+	sys_write32(mask, ICENABLER(base, idx));
+	gic_wait_rwp(intid);
+
+	/* PRIORITYR registers provide byte access */
+	sys_write8(prio & 0xff, IPRIORITYR(base, intid));
+
+	/* Interrupt type config */
+	idx = intid / 16;
+	shift = (intid & 0xF) * 2;
+
+	val = sys_read32(ICFGR(base, idx));
+	val &= ~(GICD_ICFGR_MASK << shift);
+	if (flags & IRQ_TYPE_EDGE) {
+		val |= (GICD_ICFGR_TYPE << shift);
+	}
+	sys_write32(val, ICFGR(base, idx));
+}
+
+void arm_gic_irq_enable(unsigned int intid)
+{
+	u32_t mask = BIT(intid & NUM_INTR_PER_REG);
+	u32_t idx = intid / 32;
+	mem_addr_t base;
+
+	base = (intid < GIC_SPI_BASE) ?
+	       (GIC_RDIST_BASE + GICR_SGIBASE_OFFSET)
+	       : GIC_DIST_BASE;
+
+	sys_write32(mask, ISENABLER(base, idx));
+}
+
+void arm_gic_irq_disable(unsigned int intid)
+{
+	u32_t mask = BIT(intid & NUM_INTR_PER_REG);
+	u32_t idx = intid / 32;
+	mem_addr_t base;
+
+	base = (intid < GIC_SPI_BASE) ?
+	       (GIC_RDIST_BASE + GICR_SGIBASE_OFFSET)
+	       : GIC_DIST_BASE;
+
+	sys_write32(~mask, ICENABLER(base, idx));
```

Comment:
Is this supposed to be `~mask`? Since this is clear enable register, do we have to directly write `mask` here?

Reply:
Thanks.  It must be `mask`   for` ICENABLER `.  Since we never do a `irq_disable(irq) ` we didn't see any impact.